yt_idefix/data_structures.py

```diff
@@ -724,3 +797,241 @@ def _validate_units_override_keys(cls, units_override):
+
+            if density_unit:
+                try:
+                    density_unit = float(density_unit)
+                except ValueError:
+                    for name in constant_names:
+                        density_unit = density_unit.replace(
+                            name, 'pluto_def_constants["%s"]' % name
+                        )
+                    density_unit = density_unit.replace("sqrt", "np.sqrt")
+                    density_unit = density_unit.replace("log", "np.log")
+                    density_unit = eval(density_unit)
+
+            if velocity_unit:
+                try:
+                    velocity_unit = float(velocity_unit)
+                except ValueError:
+                    for name in constant_names:
+                        velocity_unit = velocity_unit.replace(
+                            name, 'pluto_def_constants["%s"]' % name
+                        )
+                    velocity_unit = velocity_unit.replace("sqrt", "np.sqrt")
+                    velocity_unit = velocity_unit.replace("log", "np.log")
+                    velocity_unit = eval(velocity_unit)
+
+            if (density_unit) and (length_unit):
+                mass_unit = density_unit * length_unit**3
+
+        if not length_unit:
+            self.length_unit = self.quan(1.0, "AU")
+        else:
+            self.length_unit = self.quan(length_unit, "cm")
+        if not mass_unit:
+            if not density_unit:
+                mp = pluto_def_constants["CONST_mp"]
+                self.mass_unit = self.quan(
+                    self.quan(mp, "g/cm**3")
+                    * self.length_unit**3
+                    / self.quan(1.0, "Msun"),
+                    "Msun",
+                )
+            else:
+                self.mass_unit = self.quan(
+                    self.quan(density_unit, "g/cm**3")
+                    * self.length_unit**3
+                    / self.quan(1.0, "Msun"),
+                    "Msun",
+                )
+        else:
+            self.mass_unit = self.quan(mass_unit, "g")
+        if not velocity_unit:
+            self.velocity_unit = self.quan(1.0, "km/s")
+            self.time_unit = self.length_unit / self.velocity_unit
+        else:
+            self.velocity_unit = self.quan(velocity_unit, "cm/s")
+            self.time_unit = self.length_unit / self.velocity_unit
+        if not magnetic_unit:
+            magnetic_unit = self.quan(1.0, "gauss")
+        else:
+            self.magnetic_unit = self.quan(magnetic_unit, "gauss")
+
+        for key, unit in self.__class__.default_units.items():
+            setdefaultattr(self, key, self.quan(1, unit))
+
+    @classmethod
+    def _is_valid(cls, filename, *args, **kwargs):
+        # This accepts a filename or a set of arguments and returns True or
+        # False depending on if the file is of the type requested.
+        test1 = ("dbl.h5" in filename) or ("flt.h5" in filename)
```

Comment:
In this method it's important to return as early as possible, so prefer this style
```suggestion
        if not ("dbl.h5" in filename) or ("flt.h5" in filename):
            return False
```

Reply:
Checking node_coords and cell_coords ensure that it is not just any hdf5 file.

Reply:
the extension is a necessary condition, not a sufficient one, but we can still return early if we get any other extension !
I'm not suggesting to throw away the rest of the function, just that we should performs checks as greedily as possible and return with the first invalidated necessary condition.

Reply:
Okay then where do you suggest putting in the rest of the check?

Reply:
Let me clarify. What I'm suggesting is to change how the function is organised to adopt the following form:
```python
if not necessary_condition_1:
    return False
if not necessary_condition_2:
    return False
if not necessary_condition_3:
    return False
...
return True
```
or even simpler (and not any less optimal)
```python
if not (
    necessary_condition_1
    and necessary_condition_2
    and necessary_condition_3
):
    return False
...
return True
```
of course some checks may be more complex to express than others (in particular, opening the file with h5py), but that's the gist.